_posts/2019-12-24-release-r118/index.md

```diff
@@ -0,0 +1,207 @@
+
+Similarly as for Scala Stream Collector, this type of bad row is emitted whenever the size of the enriched event is too big for the message queue.
+It can also be emitted if the size of another kind of bad row is too big for the message queue. In this case it will be truncated and wrapped in a size violation bad row instead.
+
+<h2>2.5. Working with bad rows in Scala</h2>
+
+Scala library [snowplow-badrows](https://github.com/snowplow-incubator/snowplow-badrows/tree/release/0.1.0/) contains all the case classes
+defining the new bad rows and makes it easier to work with them. By using the same version of the library that the application producing bad rows,
+we can be sure that there won't be any incompatibility when processing them.
+
+To parse a `String` containing a bad row, one could write:
+
+{% highlight scala %}
+import com.snowplowanalytics.snowplow.badrows._
+
+import com.snowplowanalytics.iglu.core.circe.instances._
+import com.snowplowanalytics.iglu.core.SelfDescribingData
+
+import cats.implicits._
+
+import io.circe.parser
+import io.circe.syntax._
+
+def parseBadRow(jsonStr: String): Either[String, BadRow] =
+  for {
+    json <- parser.parse(jsonStr).leftMap(_.getMessage)
+    sdj <- SelfDescribingData.parse(json).leftMap(_.code)
+    badrow <- sdj.data.as[BadRow].leftMap(_.getMessage)
+  } yield badrow
+{% endhighlight %}
+
+It is then straightforward to determine the type of bad row with pattern matching and to access its fields:
+
+{% highlight scala %}
+val badRow: BadRow = ???
+
+badRow match {
+  case BadRow.SchemaViolations(_, failure, _) =>
+    val timestamp = failure.timestamp
+    doSomethingWithTimestamp(timestamp)
+
+  case BadRow.EnrichmentFailures(_, _, payload) =>
+    val maybeCountry = payload.enriched.geo_country
+    doSomethingWithCountry(maybeCountry)
+
+  case _ => ???
+}
+{% endhighlight %}
+
+<h1 id="refererparser">3. Referer parser enrichment improvement</h1>
+
+Referer parser enrichment requires a [file](https://github.com/snowplow-referer-parser/referer-parser/blob/develop/resources/referers.yml)
+with a list of referers to do its job.
+
+From this release, the file doesn't need to be embedded any more and its URI can be specified
+as an [input parameter](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow/referer_parser/jsonschema/2-0-0#L33-L36)
+of the enrichment.
+
+The database can thus be updated without any modification to the code.
+
+<h1 id="upgrading">4. Upgrading</h1>
+
+The upgrading guide can be found on our [wiki page](https://github.com/snowplow/snowplow/wiki/Upgrade-Guide#r118).
+
+<h1 id="roadmap">5. Roadmap</h1>
+
+While R118 contains updated versions of components emitting bad rows (Scala Stream Collector and enrich jobs),
+R119 will complete the ecosystem with the update of the components processing these bad rows:
```

Comment:
But these components are independent from snowplow/snowplow releases (R119). There's a chance we won't have R119 and all components will be separated.